### PR TITLE
feat: implement automatic socket disconnect/reconnect

### PR DESCRIPTION
**Pull request by connorlindsey**
Watches for document visibility changes and automatically disconnects and reconnects the web socket, after a delay.

### CODE REVIEW

File: packages/client/src/api.ts [hunk 1]
```diff
@@ -82,6 +82,18 @@ class ApiClient {
     }
   }
 
+  reconnectSocket() {
+    if (this.socket && !this.socket.isConnected()) {
+      this.socket.connect();
```
Comment (connorlindsey) [inline packages/client/src/api.ts:87]: TIL Phoenix sockets are stateful, so reconnecting handles rejoining channels as well

**Review by kylemcd — APPROVED** [review]

File: packages/client/src/clients/feed/feed.ts [hunk 2]
```diff
@@ -667,6 +674,42 @@ class Feed {
       console.warn(`Could not broadcast ${type}, got error: ${e}`);
     }
   }
+
+  /**
+   * Listen for changes to document visibility and automatically disconnect
+   * or reconnect the socket after a delay
+   */
+  private setupAutoSocketManager(
+    auto_manage_socket_connection: boolean | number,
```
Comment (cjbell) [inline packages/client/src/clients/feed/feed.ts:683]: Nit: I know we have this as snake cased as the option name, but internally here I would expect it as camelCase

File: packages/client/src/clients/feed/interfaces.ts [hunk 3]
```diff
@@ -22,6 +22,10 @@ export interface FeedClientOptions {
   trigger_data?: GenericData;
   // Optionally enable cross browser feed updates for this feed
   __experimentalCrossBrowserUpdates?: boolean;
+  // Optionally automatically manage socket connections on changes to tab visibility
+  // Defaults to false. If true, defaults to disconnecting after 2000ms when the tab is hidden
+  // Accepts a number as a custom delay duration in milliseconds
+  auto_manage_socket_connection?: boolean | number;
```
Comment (cjbell) [inline packages/client/src/clients/feed/interfaces.ts:28]: 🤔 I wonder if this should be 2 options, one for enabling the behavior and one for setting the disconnect window.

**Review by cjbell — COMMENTED** [review]
two tiny comments but looks great

File: packages/client/src/clients/feed/interfaces.ts (hunk 3 above)
Comment (connorlindsey) [reply packages/client/src/clients/feed/interfaces.ts:28]: Yeah, was debating that. Having one option could be a little confusing and requires more explanation. Having two options is more verbose and requires some extra logic to handle both cases (not a problem, just something to consider if e.g. a duration is set but the boolean flag is off)

Comment (connorlindsey) [reply packages/client/src/clients/feed/interfaces.ts:28]: Going to refactor to have two options. I think it's easier to understand

**Review by cjbell — APPROVED** [review]
💯💯💯 beautiful

**Pull request merged**